Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -1266,12 +1240,34 @@ extension SwiftLanguageServer {
     return Array(hints)
   }
 
+  public func syntacticDiagnosticFromBuiltInSwiftSyntax(
+    for snapshot: DocumentSnapshot
+  ) async throws -> DocumentDiagnosticReport {
+    let syntaxTree = await syntaxTreeManager.syntaxTree(for: snapshot)
+    let swiftSyntaxDiagnostics = ParseDiagnosticsGenerator.diagnostics(for: syntaxTree)
+    let diagnostics = swiftSyntaxDiagnostics.compactMap { Diagnostic($0, in: snapshot) }
+    return .full(RelatedFullDocumentDiagnosticReport(items: diagnostics))
+  }
+
   public func documentDiagnostic(_ req: DocumentDiagnosticsRequest) async throws -> DocumentDiagnosticReport {
     guard let snapshot = documentManager.latestSnapshot(req.textDocument.uri) else {
       let msg = "failed to find snapshot for url \(req.textDocument.uri)"
       log(msg)
       throw ResponseError.unknown(msg)
     }
+    guard let buildSettings = await self.buildSettings(for: req.textDocument.uri), !buildSettings.isFallback else {
+      log(
+        "Producing syntactic diagnostics from the built-in swift-syntax because we have fallback arguments",
+        level: .info
+      )
+      // If we don't have build settings or we only have fallback build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      return try await syntacticDiagnosticFromBuiltInSwiftSyntax(for: snapshot)
+    }
```

Comment:
This is cool. One thought though - sourcekitd will ignore a bunch of diagnostics (eg. editor placeholder). We have to dupe that here?

Reply:
Good point. The editor placeholder diagnostic is the only one I can think of though. Do you know of any others that sourcekitd ignores?

Reply:
The two I see in SwiftEditor are:
```
  if (Info.ID == diag::lex_editor_placeholder.ID ||
      Info.ID == diag::oslog_invalid_log_message.ID)
```

And the latter is a SIL, so not a problem here.